Add BaseBuilderBotModule and BuildingRepairBotModule

reaperrr opened the pull request:
Extracts all logic related to base-building from `HackyAI` into a `BaseBuilderBotModule`. `BaseBuilder` is kept as separate .cs file, but of course adapted.

Additionally, replaced the `ShouldRepairBuildings` boolean with a separate `RepairBuildingsBotModule`, as some singleplayer missions may want only that.

Depends on #15786.

reaperrr committed:
Extract BaseBuilderBotModule from HackyAI

reaperrr commented:
Ping @IceReaper. Would be great if you could first look at #15786 (unless it got merged in the meantime), since that's a dependency and will make this PR at least a little more readable as well.

pchote commented on OpenRA.Mods.Common/AI/HackyAI.cs:
```diff
@@ -227,19 +160,22 @@ static object LoadBuildingCategories(MiniYaml yaml)
 		public readonly HackyAIInfo Info;
 
-		public CPos GetRandomBaseCenter()
+		public CPos GetRandomBaseCenter
```
Style nit: this would be best kept as a method. It's doing a calculation and returning a different result each time, whereas properties are generally used for state-like data.

pchote commented on OpenRA.Mods.Common/AI/HackyAI.cs:
```diff
@@ -371,49 +293,25 @@ ActorInfo ChooseUnitToBuild(ProductionQueue queue)
 		}
 
-		IEnumerable<Actor> GetActorsWithTrait<T>()
+		bool HasAdequateConstructionYardCount
```
These ones return consistent results (assuming other state doesn't change), so are appropriate to leave as properties.

pchote commented on OpenRA.Mods.Common/AI/HackyAI.cs:
```diff
@@ -939,37 +762,30 @@ void BuildUnit(string category, string name)
+			// TODO: Add an option to include this in CheckSyncUnchanged.
+			// Checking sync for this is too expensive to include it by default,
+			// so it should be implemented as separate sub-option checkbox.
+			using (new PerfSample("attackResponse_bots"))
```
Style nit: I don't think we use camelcase anywhere else in perf sample names.  It might be better to rename this to `bot_attack_response` and `tick_bots` to `bot_tick`?

reaperrr replied:
Done.

pchote commented on OpenRA.Mods.Common/Traits/BotModules/BaseBuilderBotModule.cs:
```diff
@@ -0,0 +1,246 @@
+
+	public class BaseBuilderBotModule : ConditionalTrait<BaseBuilderBotModuleInfo>, IBotTick, IBotPositionsUpdated, IBotRespondToAttack
+	{
+		public CPos GetRandomBaseCenter
```
As in HackyAI, this should be a method.

pchote commented on OpenRA.Mods.Common/Traits/BotModules/BuildingRepairBotModule.cs:
```diff
@@ -0,0 +1,42 @@
+		public BuildingRepairBotModule(Actor self, BuildingRepairBotModuleInfo info)
+			: base(info) { }
+
+		void IBotRespondToAttack.RespondToAttack(IBot bot, Actor self, AttackInfo e)
```
Do we need to worry about repairs being automatically cancelled by something else?

reaperrr replied:
No idea.
But unless I'm missing something, this shouldn't break anything that wasn't already broken on bleed, at least.

pchote commented on OpenRA.Mods.Common/Traits/BotModules/BuildingRepairBotModule.cs:
```diff
@@ -0,0 +1,42 @@
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Manages AI repairing base buildings.")]
+	public class BuildingRepairBotModuleInfo : ConditionalTraitInfo
```
It would be cool to expand this in the future with options for starting/stopping repairs based on current funds, and repairing ally structures.

reaperrr replied:
And customizable DamageState thresholds, maybe even per BuildingCommonName (indirectly giving priority to some building types over others).

pchote commented on mods/cnc/rules/ai.yaml:
```diff
@@ -289,3 +184,140 @@ Player:
+		BuildingFractions:
+			proc: 17%
+			pyle: 2%
+			hand: 2%
```
The automatic % &rarr; float conversions isn't used anywhere else, IIRC.  This would be a good opportunity to switch these over to a different syntax.

reaperrr replied:
Do you mean something like integer percentages, or something different entirely?

reaperrr commented:
Oh wow, I broke unit production somehow...

reaperrr commented:
Updated.

Note: There's a weird bug in the update rule I haven't been able to figure out yet. The `BaseBuilderBotModule` of the first HackyAI in a yaml (`rush` in RA, for example) somehow gets the same `RequiredCondition` setup as the default `HarvesterBotModule` (fixed that manually for now). It works fine for the other AIs and I can't get my head around when, where and why it does that.

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/20180923/ExtractHackyAIModules.cs:
```diff
@@ -124,28 +210,66 @@ public override IEnumerable<string> UpdateActorNode(ModData modData, MiniYamlNod
+							continue;
 
-					addNodes.Add(spNode);
+						requiresConditionNode.ReplaceValue(oldValue + " || " + conditionString);
```
You are adding the same `requiresCondition` to all of the module type, and yaml nodes are a reference type so when you change this you actually change *all* nodes!

The fix is to create the `requiresCondition` fresh for each module you wish to add it to.

pchote commented on OpenRA.Mods.Common/Traits/BotModules/BaseBuilderBotModule.cs:
```diff
@@ -0,0 +1,243 @@
+
+		[Desc("Tells the AI what building types fall under the same common name.",
+			"Possible keys are ConstructionYard, Power, Refinery, Silo, Barracks, Production, VehiclesFactory, NavalProduction.")]
+		[FieldLoader.LoadUsing("LoadBuildingCategories", true)]
```
I'm not sure we win anything by keeping these as a sub-dictionary on the new module.  Can we define `VehiclesFactoryTypes` and so on directly?

pchote commented:
Needs a rebase now too

reaperrr committed:
Fix bot module update rule setting wrong RequiresCondition

Yaml nodes are reference types, so caching this meant changes would be applied on all of them.
Additionally, only add HarvesterBotModule if at least one AI is actually using it.

reaperrr committed:
BaseBuilder- and BuildingRepairBotModule update rule

obrakmann commented:
Looks ok to me, didn't notice any regressions while observing an all-AI match. :+1:

pchote reviewed (approved):
LGTM in the context of moving existing code around to enable future cleanups. There is a obviously a lot I would like to see change in this code, but this isn't the time or place to start.

Lets see what regressions fall out once more players get to test this...

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/a6825ebdad6f0d94ea130759c573bbf831a6b195
